File: CrackCode/movezeroes.c
/* move all 0s to the front of an integer array */
#include <stdio.h>

int 
move (int *arr, int n) {
	int i = 0, start = 0, tmp=0;
	for (i=0; i<n; i++) {
		if (arr[i] == 0) {
			if (arr[start] != 0) {
				tmp = arr[start];
				arr[start] = arr[i];
				arr[i] = tmp;
			}
			start++;
		}
	}
	printf ("New array: ");
	for (i=0; i<n; i++) {
		printf ("%d ", arr[i]);
	}
	printf ("\n");
	return n-start;
}

int main ()
{
	int arr1[] = {5, 0, 2, 5, 0};
	int arr2[] = {5, 5, 2, 5, 5};
	int arr3[] = {0, 0, 0, 0, 0};
	printf ("%d.\n", move(arr1, sizeof(arr1)/sizeof(arr1[0])));
	printf ("%d.\n", move(arr2, sizeof(arr2)/sizeof(arr2[0])));
	printf ("%d.\n", move(arr3, sizeof(arr3)/sizeof(arr3[0])));
	return 0;
}

File: CrackCode/JAVA/MoveZeroes.java
public class MoveZeroes {
  static void move(int[] array) {
    int len = array.length;
    int end = len-1;  
    int temp;
    
    for (int i=0; i<end;) {
      if (array[i] == 0) {
        temp = array[end];
        array[end] = 0;
        array[i] = temp;
        end--;
      } else {
        i++;
      }
    }
  }
  
  static void move2(int[] array) {
    int curr = 0; int len = array.length;

    for (int i=0; i< len; i++) {
      if (array[i] != 0) {
        array[curr++] = array[i];
      }
    }

   for (; curr<len; curr++) {
     array[curr]=0;
   }

  }
        

  public static void main (String args[]) {
    int[] arr = {0, 1, 3, 0, 5, 3, 0, 7};
    move2(arr);
    for (int i = 0; i < arr.length; i++) {
      System.out.println(arr[i] + ", ");
    }
  }
}
